### ISSUE TITLE
Bundle size

### ISSUE THREAD

**Issue by ritz078** [bvaughn/react-virtualized#427]
Hi @bvaughn ,

We are glad that this library is performing so well on mobile devices for us. One thing that we are concerned about is the bundle size. We are using webpack 2 for bundling with tree-shaking. Currently react-virtualized's size in our bundle can be seen below.

<img width="816" alt="screen shot 2016-10-07 at 7 38 21 pm" src="https://cloud.githubusercontent.com/assets/5389035/19200444/a5020516-8ce6-11e6-82f9-ea9b0ec77ba1.png">

This is when we are using `InfiniteScroller` with `List`. Even though I am fine with this, I would like to know whether this can be improved ? Because even slight reductions in file size can do wonders on mobile.

**Comment by bvaughn** [comment]
Hi @ritz078. Nice to hear from you! 😄

That is a sweet looking breakdown. Help me better understand what it means though. For example, why does it show the `dist/commonjs` folder? Shouldn't you be using `dist/es` if you want the tree-shaking benefits of Webpack 2? (That's what the `jsnext:main` target points at.)

That should slim the module down a bit more for you as it wouldn't pull in any of the HOCs besides `InfiniteScroller` (which is small) or any of the `Collection` or `Table` related code.

**Comment by ritz078** [comment]
Yes you are right. it should have taken files from `jsnext:main`. I checked again and the  files are loading from `dist` directory. Even checked the version of webpack in stats.json just to be sure 😄 . Seems to be an issue with webpack ? Either that's webpack issue or something in our configuration. 

cc @TheLarkInn Can any error in configuration cause this error where the files are not taken from `jsnext:main` ?

**Comment by TheLarkInn** [comment]
Only latest webpack pulls from the "module" field not jsnext:main. Otherwise you have to specify a seperate field in your config under resolver.modules

**Comment by bvaughn** [comment]
You could also import from `dist/es` but that's a little janky.

```js
import { List } from 'react-virtualized/dist/es';
```

**Comment by ritz078** [comment]
@bvaughn I guess that means "module" works for webpack 2 and "jsnext:main" for rollup.

**Comment by bvaughn** [comment]
Oh I see! Seems related to webpack/webpack/issues/1979

I don't mind doing a quick bugfix release to add the "module" attribute to `package.json` if that's the preferred route now. I was unaware.

**Comment by ritz078** [comment]
was going through that PR only. [Seems like even rollup is now supporting 'module'](https://github.com/rollup/rollup-plugin-node-resolve/pull/49).

**Comment by bvaughn** [comment]
@ritz078 git 8.0.12 a quick try and let me know if the results are improved for you.

Also, thanks for the quick reply @TheLarkInn! Very helpful.

**Comment by bvaughn** [comment]
Yeah, looks like I wasn't following along closely enough with the [new lookup rules](https://github.com/dherman/defense-of-dot-js/blob/master/proposal.md#detailed-lookup-rules)

**Comment by ritz078** [comment]
Here's the new contribution of react-virtualized.

<img width="853" alt="screen shot 2016-10-08 at 3 36 43 am" src="https://cloud.githubusercontent.com/assets/5389035/19206714/98023da0-8d08-11e6-83be-50d84237152f.png">

but the chunk's size has increased by 18 KB. 😞

**Comment by bvaughn** [comment]
Ok. That's looking better in terms of `dist/es`.

My next question would be- why do `Table` and `Collection` appear in the chart? If you're only using `List` then you should have `List` + `Grid` only

**Comment by ritz078** [comment]
yes that's my doubt too. 

the code is simply

```js
import { InfiniteLoader, AutoSizer, List } from 'react-virtualized'
```

**Comment by ritz078** [comment]
So it seems that tree shaking isn't working. If I change the above code to

```js
import { InfiniteLoader } from 'react-virtualized/dist/es/InfiniteLoader'
import { AutoSizer } from 'react-virtualized/dist/es/AutoSizer'
import { List } from 'react-virtualized/dist/es/List'
```

file size change significantly. huge win this way.

<img width="945" alt="screen shot 2016-10-08 at 4 22 17 am" src="https://cloud.githubusercontent.com/assets/5389035/19207490/1da0d588-8d0f-11e6-8849-f6e2d9a2c99e.png">

cc: @TheLarkInn

**Comment by bvaughn** [comment]
Maybe my [`dist/es/index.js`](https://unpkg.com/react-virtualized@8.0.12/dist/es/index.js) is stupid?

**Comment by ritz078** [comment]
Its not working that way. Even individually picking `es` build was giving error during execution. Finally I am loading `commonjs` versions individually.

**Comment by bvaughn** [comment]
Could you elaborate a bit more @ritz078? What wasn't working about the es build?

**Comment by ritz078** [comment]
Here are all the cases that I tried

**Case I**

```js
import {InfiniteLoader, AutoSizer, List} from 'react-virtualized'
```
Build: ✅ 
Loading form : `es`
File size: **increases**
Tree shaking: ❌ 

**Case II**
```js
import { InfiniteLoader } from 'react-virtualized/InfiniteLoader'
import { AutoSizer } from 'react-virtualized/AutoSizer'
import { List } from 'react-virtualized/List'
```
Build: ❌ can't resolve modules

**Case III**
```js
import { InfiniteLoader } from 'react-virtualized/dist/es/InfiniteLoader'
import { AutoSizer } from 'react-virtualized/dist/es/AutoSizer'
import { List } from 'react-virtualized/dist/es/List'
```
Build: ✅ 
File size decreases
But when I make a request, it gives an error `"Unhandled Promise rejection: SyntaxError: Unexpected token import"`. This has nothing to do with react-virtualized. We even tried this with react-router and got the same error. So either webpack issue or our config. Its really tough with the lack of docs on webpack 2 currently.

**Case IV**

```js
import { InfiniteLoader } from 'react-virtualized/dist/commonjs/InfiniteLoader'
import { AutoSizer } from 'react-virtualized/dist/commonjs/AutoSizer'
import { List } from 'react-virtualized/dist/commonjs/List'
```
Build: ✅ 
File size : **decreases**
Everything works fine. 

So currently we are using this. Hopefully we will be able to figure out the webpack issue soon.

**Comment by bvaughn** [comment]
I see. The error you mentioned (_"Unhandled Promise rejection: SyntaxError: Unexpected token import"_) sounds like Webpack is configured not to transpile modules within the `node_modules` folder- which is normally a reasonable setting, although in this case... it doesn't make sense. Not sure what the appropriate approach is w.r.t. es modules.

**Comment by ritz078** [comment]
Hopefully we will be able to figure that out soon. Closing this issue. Thanks for the help.

**Comment by bvaughn** [comment]
Any time @ritz078!

Please share any additional findings if you think of it. 😁

**Comment by ritz078** [comment]
Surely.

**Comment by SpencerCarstens** [comment]
Ran into the same exact issue. Webpack 2 and tree-shaking did not work until:

```javascript
// Changing...
import { AutoSizer, Column, Table, InfiniteLoader } from 'react-virtualized';

// To...
import { InfiniteLoader } from 'react-virtualized/dist/es/InfiniteLoader';
import { AutoSizer } from 'react-virtualized/dist/es/AutoSizer';
import { Column, Table } from 'react-virtualized/dist/es/Table';
```

👍

**Comment by rt2zz** [comment]
thanks for the solutions, works for my setup

@ritz078 what bundle size analyzer are you using to get those graphs?

**Comment by rubencodes** [comment]
Same, halved the size of my `react-virtualized` imports. Thanks for sharing!

**Comment by romulof** [comment]
I've done some investigation on this same problem today.
Here are my findings on why importing everything from index bugs webpack tree shaking:

1. The syntax `export NS from`/`export default from` still a proposal. It ain't ES6/ES2015 standard.
2. The most simple way to transform this notation is calling a standard `import`  then `export` the imported module.
3. webpack2 Tree Shaking works by removing the `export` statement, so the `import` statement is kept.

Take a look at webpack output for my project (without mangling and keeping the comments):
```es6
/* exports provided: ArrowKeyStepper, AutoSizer, CellMeasurer, CellMeasurerCache, Collection, ColumnSizer, accessibilityOverscanIndicesGetter, defaultCellRangeRenderer, defaultOverscanIndicesGetter, Grid, InfiniteLoader, List, createMasonryCellPositioner, Masonry, MultiGrid, ScrollSync, defaultTableCellDataGetter, defaultTableCellRenderer, defaultTableHeaderRenderer, defaultTableHeaderRowRenderer, defaultTableRowRenderer, Table, Column, SortDirection, SortIndicator, WindowScroller */
/* exports used: CellMeasurerCache, CellMeasurer, AutoSizer, InfiniteLoader, List */
/*!**********************************************!*\
  !*** ./~/react-virtualized/dist/es/index.js ***!
  \**********************************************/
/***/
function(module, __webpack_exports__, __webpack_require__) {
    "use strict";
    /* harmony import */
    var __WEBPACK_IMPORTED_MODULE_1__AutoSizer__ = (__webpack_require__(/*! ./ArrowKeyStepper */ 847), 
    __webpack_require__(/*! ./AutoSizer */ 848));
    /* harmony reexport (binding) */
    __webpack_require__.d(__webpack_exports__, "c", function() {
        return __WEBPACK_IMPORTED_MODULE_1__AutoSizer__.a;
    });
    /* harmony import */
    var __WEBPACK_IMPORTED_MODULE_2__CellMeasurer__ = __webpack_require__(/*! ./CellMeasurer */ 353);
    /* harmony reexport (binding) */
    __webpack_require__.d(__webpack_exports__, "b", function() {
        return __WEBPACK_IMPORTED_MODULE_2__CellMeasurer__.a;
    }), /* harmony reexport (binding) */
    __webpack_require__.d(__webpack_exports__, "a", function() {
        return __WEBPACK_IMPORTED_MODULE_2__CellMeasurer__.b;
    });
    /* harmony import */
    var __WEBPACK_IMPORTED_MODULE_6__InfiniteLoader__ = (__webpack_require__(/*! ./Collection */ 853), 
    __webpack_require__(/*! ./ColumnSizer */ 855), __webpack_require__(/*! ./Grid */ 147), 
    __webpack_require__(/*! ./InfiniteLoader */ 861));
    /* harmony reexport (binding) */
    __webpack_require__.d(__webpack_exports__, "d", function() {
        return __WEBPACK_IMPORTED_MODULE_6__InfiniteLoader__.a;
    });
    /* harmony import */
    var __WEBPACK_IMPORTED_MODULE_7__List__ = __webpack_require__(/*! ./List */ 862);
    /* harmony reexport (binding) */
    __webpack_require__.d(__webpack_exports__, "e", function() {
        return __WEBPACK_IMPORTED_MODULE_7__List__.a;
    });
    /* harmony import */
    __webpack_require__(/*! ./Masonry */ 865), __webpack_require__(/*! ./MultiGrid */ 867), 
    __webpack_require__(/*! ./ScrollSync */ 868), __webpack_require__(/*! ./Table */ 869), 
    __webpack_require__(/*! ./WindowScroller */ 870);
}, /* 873 */
```

The comments on the first 2 lines shows exactly that tree shaking algorithm is working, and you can see at the end of the module that the harmony imports of the unused modules are still there, but no exports.

I tried removing `babel-plugin-transform-export-extensions` to deliver the whole `export ns from` syntax to webpack, allowing it to properly manage the unused `export {} from`, but It didn't change the result.

My conclusion is that this is actually a webpack issue or an unimplemented feature. For now it's better to use direct imports like `import { List } from 'react-virtualized/dist/es/List'`.

Right now I'm trying to find something related on webpack issues.

**Comment by drKnoxy** [comment]
@TheLarkInn any chance you could weigh in on where the issue should be corrected; webpack / library

The issue with the es index.js file's export syntax doesn't seem to allow for tree shaking
```js
export { ArrowKeyStepper } from "./ArrowKeyStepper";
export { AutoSizer } from "./AutoSizer";
export { CellMeasurer, CellMeasurerCache } from "./CellMeasurer";
export { Collection } from "./Collection";
export { ColumnSizer } from "./ColumnSizer";
export { accessibilityOverscanIndicesGetter, defaultCellRangeRenderer, defaultOverscanIndicesGetter, Grid } from "./Grid";
export { InfiniteLoader } from "./InfiniteLoader";
export { List } from "./List";
export { createCellPositioner as createMasonryCellPositioner, Masonry } from "./Masonry";
export { MultiGrid } from "./MultiGrid";
export { ScrollSync } from "./ScrollSync";
export { defaultCellDataGetter as defaultTableCellDataGetter, defaultCellRenderer as defaultTableCellRenderer, defaultHeaderRenderer as defaultTableHeaderRenderer, defaultHeaderRowRenderer as defaultTableHeaderRowRenderer, defaultRowRenderer as defaultTableRowRenderer, Table, Column, SortDirection, SortIndicator } from "./Table";
export { WindowScroller } from "./WindowScroller";
```

**Comment by TrySound** [comment]
@drKnoxy The problem is untreeshakable classes which are transpilled to function constructor with prototype. It's side effect for bundlers. All component based libraries affected by this. Just use paths like `import AutoSizer from 'react-virtualized/dist/es/AutoSizer';`. In the future `side-effects` field may fix this.

**Comment by drKnoxy** [comment]
Doesn't that sound really lame though? Feels feature incomplete... for
either treeshaking / import syntax / uglify

**Comment by TrySound** [comment]
It works, but it's not panacea. Static analyzing is hard when you deal with javascript.

**Comment by alexparish** [comment]
@TrySound Now the `side-effects` fields has been added to package.json should it be possible to use the following import and expect webpack 4 to tree-shake the imports?

`import {InfiniteLoader, AutoSizer, List} from 'react-virtualized'

I tried this but my bundle did not decrease in size.

**Comment by alexparish** [comment]
@TrySound That would be great. 

I believe the only impediment to tree-shaking the raw source code of this library is caused by the addition of the `babel` property in package.json.

Having added `jsnext:main` to the mainFields property of my webpack config in order to pick up the source from `dist/es/index.js`, Babel is trying to find `./.babelrc.js` in the project but failing:

```
Module build failed: Error: Couldn't find preset "./.babelrc.js" relative to directory "/Users/parialea/Projects/dbk-ui/node_modules/react-virtualized"
    at Array.map (<anonymous>)
```

Out of interest, for v10 would you make the commonjs output tree-shakable? Is that even possible?

**Comment by TrySound** [comment]
@alexparish ignore node_modules
```js
{
  test: /\.js$/,
  exclude: /node_modules/,
  loader: 'babel-loader'
}
```

**Comment by TrySound** [comment]
And no, commonjs is not treeshakable. However rollup plugin commonjs may on some cases convert to treeshakable esm.

**Comment by TrySound** [comment]
@miguelocarvajal If there is such problem you probably overtranspile your vendors. Is it fast?

**Comment by miguelocarvajal** [comment]
@TrySound Fast as it has ever been :)

To make sure its not my configuration I just tried creating a new project with react-native and I still get the same message. These are the steps I followed:

1. react-native init VirtualizedTest
2. yarn add react-virtualized
3. Modified App.js to include an empty InfiniteLoader
4. react-native run-ios

Just with those steps and now adding any other dependencies the app no longer runs due to that babel config.
